service/src/java/org/apache/hive/service/auth/jwt/JWTValidator.java

```diff
@@ -0,0 +1,94 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hive.service.auth.jwt;
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.JWSObject;
+import com.nimbusds.jose.JWSVerifier;
+import com.nimbusds.jose.crypto.factories.DefaultJWSVerifierFactory;
+import com.nimbusds.jose.jwk.AsymmetricJWK;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.security.sasl.AuthenticationException;
+import java.io.IOException;
+import java.security.Key;
+import java.text.ParseException;
+import java.util.Date;
+import java.util.List;
+
+public class JWTValidator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JWTValidator.class.getName());
+  private final URLBasedJWKSProvider jwksProvider;
+  private static final DefaultJWSVerifierFactory verifierFactory = new DefaultJWSVerifierFactory();
+
+  public JWTValidator(HiveConf conf) throws IOException, ParseException {
+    this.jwksProvider = new URLBasedJWKSProvider(conf);
+  }
+
+  public String validateJWTAndExtractUser(String signedJwt) throws ParseException, AuthenticationException {
+    final SignedJWT parsedJwt = SignedJWT.parse(signedJwt);
+    List<JWK> matchedJWKS = jwksProvider.getJWKs(parsedJwt.getHeader());
+
+    // verify signature
+    for (JWK matchedJWK : matchedJWKS) {
+      try {
+        JWSVerifier verifier = getVerifier(parsedJwt.getHeader(), matchedJWK);
+        if (parsedJwt.verify(verifier)) {
+          break;
+        }
+      } catch (JOSEException e) {
+        LOG.info("Failed to verify JWT {} by JWK {} because {}", parsedJwt.getHeader(), matchedJWK.getKeyID(),
+            e.getMessage());
+      }
+    }
+    if (parsedJwt.getState() != JWSObject.State.VERIFIED) {
+      throw new AuthenticationException("Failed to verify JWT signature");
+    }
+
+    // verify claims
+    JWTClaimsSet claimsSet = parsedJwt.getJWTClaimsSet();
+    Date expirationTime = claimsSet.getExpirationTime();
+    if (expirationTime != null) {
+      Date now = new Date();
+      if (now.after(expirationTime)) {
+        throw new AuthenticationException("JWT has been expired");
+      }
+    }
+
+    // We assume the subject of claims is the query user
+    return claimsSet.getSubject();
+  }
+
+  private static JWSVerifier getVerifier(JWSHeader header, JWK jwk) throws JOSEException {
+    Key key = null;
+    if (jwk instanceof AsymmetricJWK) {
```

Comment:
@sourabh912 
1. This is for security. I think we should use asymmetric algorithm (public key) to verify a token, instead of secret key. Using a secret key to verify means anybody can get key and sign the JWT.
2. This method is not specific to a instance, but I checked again the library and realized that `Verifier` only depends on the key. With that being said, maybe we can put them in a map and reuse them.
Let me know if these don't make sense.

Reply:
I think the application (in this case HS2) should not enforce that **only** asymmetric algorithm is allowed though we can say that asymmetric is preferable. I don't see a very valid reason of supporting only asymmetric algo for token verification unless I am missing something.